# minst hand written digits recognition 

In [37]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist 
from keras.models import Sequential
from keras.layers.core import Dense , Activation,Dropout
from keras.optimizers import SGD,RMSprop , Adam
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility

#### load dataset

In [15]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

#### reshape the dataset to make images flat


In [16]:
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

#### noramlize images

In [17]:
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

60000 train samples
10000 test samples


#### let's define our model hyper paramters

In [14]:
NB_EPOCH = 200
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10   # number of outputs = number of digits
OPTIMIZER = SGD() # SGD optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION

#### convert target values to OHE 

In [18]:
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)

#### define our model architecture

In [19]:
model = Sequential()
model.add(Dense(NB_CLASSES, input_shape=(RESHAPED,)))
model.add(Activation('softmax'))


In [20]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_2 (Dense)              (None, 10)                7850      
_________________________________________________________________
activation_2 (Activation)    (None, 10)                0         
Total params: 7,850
Trainable params: 7,850
Non-trainable params: 0
_________________________________________________________________


#### compile our keras model 

In [21]:
model.compile(loss='categorical_crossentropy',
              optimizer=OPTIMIZER,
              metrics=['accuracy'])

In [22]:
history = model.fit(X_train, Y_train,
                    batch_size=BATCH_SIZE, epochs=NB_EPOCH,
                    verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 1s 15us/step - loss: 1.3633 - acc: 0.6796 - val_loss: 0.8904 - val_acc: 0.8246
Epoch 2/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.7913 - acc: 0.8272 - val_loss: 0.6572 - val_acc: 0.8546
Epoch 3/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.6436 - acc: 0.8497 - val_loss: 0.5625 - val_acc: 0.8681
Epoch 4/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.5717 - acc: 0.8602 - val_loss: 0.5098 - val_acc: 0.8765
Epoch 5/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.5276 - acc: 0.8678 - val_loss: 0.4758 - val_acc: 0.8826
Epoch 6/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.4973 - acc: 0.8726 - val_loss: 0.4515 - val_acc: 0.8866
Epoch 7/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.4748 - acc: 0.8775 - val_loss: 0.4333 - 

48000/48000 [==============================] - 1s 12us/step - loss: 0.3143 - acc: 0.9128 - val_loss: 0.3022 - val_acc: 0.9151
Epoch 60/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.3136 - acc: 0.9129 - val_loss: 0.3016 - val_acc: 0.9161
Epoch 61/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.3130 - acc: 0.9133 - val_loss: 0.3011 - val_acc: 0.9158
Epoch 62/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.3123 - acc: 0.9131 - val_loss: 0.3007 - val_acc: 0.9151
Epoch 63/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.3117 - acc: 0.9136 - val_loss: 0.3003 - val_acc: 0.9156
Epoch 64/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.3110 - acc: 0.9137 - val_loss: 0.2997 - val_acc: 0.9158
Epoch 65/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.3105 - acc: 0.9137 - val_loss: 0.2992 - val_acc: 0.9159
Epoch 66/200
48000/48000 [==============

Epoch 118/200
48000/48000 [==============================] - 1s 13us/step - loss: 0.2902 - acc: 0.9192 - val_loss: 0.2846 - val_acc: 0.9212
Epoch 119/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2900 - acc: 0.9191 - val_loss: 0.2844 - val_acc: 0.9212
Epoch 120/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2898 - acc: 0.9192 - val_loss: 0.2842 - val_acc: 0.9212
Epoch 121/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2895 - acc: 0.9191 - val_loss: 0.2841 - val_acc: 0.9212
Epoch 122/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2892 - acc: 0.9192 - val_loss: 0.2840 - val_acc: 0.9212
Epoch 123/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2890 - acc: 0.9194 - val_loss: 0.2838 - val_acc: 0.9211
Epoch 124/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2888 - acc: 0.9197 - val_loss: 0.2837 - val_acc: 0.9210
Epoch 125/200
48000/

48000/48000 [==============================] - 1s 12us/step - loss: 0.2793 - acc: 0.9220 - val_loss: 0.2775 - val_acc: 0.9232
Epoch 177/200
48000/48000 [==============================] - 1s 13us/step - loss: 0.2791 - acc: 0.9219 - val_loss: 0.2774 - val_acc: 0.9234
Epoch 178/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2790 - acc: 0.9221 - val_loss: 0.2774 - val_acc: 0.9228
Epoch 179/200
48000/48000 [==============================] - 1s 13us/step - loss: 0.2788 - acc: 0.9221 - val_loss: 0.2773 - val_acc: 0.9232
Epoch 180/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2787 - acc: 0.9221 - val_loss: 0.2771 - val_acc: 0.9235
Epoch 181/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2785 - acc: 0.9223 - val_loss: 0.2770 - val_acc: 0.9232
Epoch 182/200
48000/48000 [==============================] - 1s 12us/step - loss: 0.2784 - acc: 0.9220 - val_loss: 0.2769 - val_acc: 0.9231
Epoch 183/200
48000/48000 [=======

#### evaluate our model

In [23]:
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("\nTest score:", score[0])
print('Test accuracy:', score[1])

10000/10000 [==============================] - 0s 15us/step

Test score: 0.2773858499318361
Test accuracy: 0.9227


# let's improve this model by adding more hidden layers

In [27]:
### i will modifiy model architecture only
model = Sequential()
model.add(Dense(128,input_shape=(784,)))
model.add(Activation('relu'))
model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dense(10))
model.add(Activation('softmax'))

model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_6 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_7 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_7 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 10)                1290      
_________________________________________________________________
activation_8 (Activation)    (None, 10)                0         
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________


In [28]:
model.compile(loss='categorical_crossentropy',optimizer=OPTIMIZER,metrics=['accuracy'])
model.fit(X_train,Y_train,epochs=NB_EPOCH,
          batch_size=BATCH_SIZE,verbose=True,
          validation_split=0.2)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 1s 26us/step - loss: 1.5150 - acc: 0.6191 - val_loss: 0.7662 - val_acc: 0.8373
Epoch 2/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.6066 - acc: 0.8507 - val_loss: 0.4564 - val_acc: 0.8827
Epoch 3/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.4421 - acc: 0.8797 - val_loss: 0.3766 - val_acc: 0.8963
Epoch 4/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.3821 - acc: 0.8928 - val_loss: 0.3380 - val_acc: 0.9067
Epoch 5/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.3481 - acc: 0.9014 - val_loss: 0.3144 - val_acc: 0.9106
Epoch 6/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.3248 - acc: 0.9073 - val_loss: 0.2988 - val_acc: 0.9157
Epoch 7/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.3070 - acc: 0.9116 - val_loss: 0.2832 - 

48000/48000 [==============================] - 1s 22us/step - loss: 0.0890 - acc: 0.9757 - val_loss: 0.1165 - val_acc: 0.9654
Epoch 60/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.0875 - acc: 0.9764 - val_loss: 0.1162 - val_acc: 0.9662
Epoch 61/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.0861 - acc: 0.9766 - val_loss: 0.1156 - val_acc: 0.9659
Epoch 62/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.0848 - acc: 0.9770 - val_loss: 0.1144 - val_acc: 0.9667
Epoch 63/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.0834 - acc: 0.9771 - val_loss: 0.1131 - val_acc: 0.9672
Epoch 64/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.0821 - acc: 0.9776 - val_loss: 0.1123 - val_acc: 0.9672
Epoch 65/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.0810 - acc: 0.9780 - val_loss: 0.1120 - val_acc: 0.9676
Epoch 66/200
48000/48000 [==============

48000/48000 [==============================] - 1s 24us/step - loss: 0.0406 - acc: 0.9905 - val_loss: 0.0926 - val_acc: 0.9734
Epoch 118/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.0400 - acc: 0.9904 - val_loss: 0.0921 - val_acc: 0.9733
Epoch 119/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.0396 - acc: 0.9906 - val_loss: 0.0919 - val_acc: 0.9739
Epoch 120/200
48000/48000 [==============================] - 1s 24us/step - loss: 0.0392 - acc: 0.9909 - val_loss: 0.0914 - val_acc: 0.9737
Epoch 121/200
48000/48000 [==============================] - 1s 24us/step - loss: 0.0387 - acc: 0.9909 - val_loss: 0.0915 - val_acc: 0.9737
Epoch 122/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.0382 - acc: 0.9911 - val_loss: 0.0911 - val_acc: 0.9741
Epoch 123/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.0377 - acc: 0.9911 - val_loss: 0.0917 - val_acc: 0.9738
Epoch 124/200
48000/48000 [=======

48000/48000 [==============================] - 1s 24us/step - loss: 0.0210 - acc: 0.9962 - val_loss: 0.0893 - val_acc: 0.9749
Epoch 176/200
48000/48000 [==============================] - 1s 25us/step - loss: 0.0207 - acc: 0.9962 - val_loss: 0.0899 - val_acc: 0.9742
Epoch 177/200
48000/48000 [==============================] - 1s 26us/step - loss: 0.0205 - acc: 0.9961 - val_loss: 0.0902 - val_acc: 0.9738
Epoch 178/200
48000/48000 [==============================] - 1s 23us/step - loss: 0.0203 - acc: 0.9964 - val_loss: 0.0892 - val_acc: 0.9747
Epoch 179/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.0201 - acc: 0.9964 - val_loss: 0.0893 - val_acc: 0.9743
Epoch 180/200
48000/48000 [==============================] - 1s 22us/step - loss: 0.0199 - acc: 0.9965 - val_loss: 0.0898 - val_acc: 0.9742
Epoch 181/200
48000/48000 [==============================] - 1s 24us/step - loss: 0.0197 - acc: 0.9965 - val_loss: 0.0896 - val_acc: 0.9737
Epoch 182/200
48000/48000 [=======

##### evaluation after improvments

In [30]:
score = model.evaluate(X_test,Y_test)
print("\nTest score:", score[0])
print('Test accuracy:', score[1]*100)

10000/10000 [==============================] - 0s 23us/step

Test score: 0.07980386844545137
Test accuracy: 97.6


Now we reached 97.6 % accuracy on test set . but there are more improvements ? , let's see

let's add some dropout to prevent overfitteing 

In [32]:

model = Sequential()
model.add(Dense(128,input_shape=(784,)))
model.add(Activation('relu'))
model.add(Dropout(.3)) # drop .3 of connections between layers

model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dropout(.3))

model.add(Dense(10))
model.add(Activation('softmax'))

model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_9 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_9 (Activation)    (None, 128)               0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_10 (Dense)             (None, 128)               16512     
_________________________________________________________________
activation_10 (Activation)   (None, 128)               0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_11 (Dense)             (None, 10)                1290      
__________

In [34]:
model.compile( loss='categorical_crossentropy',
              optimizer=OPTIMIZER,metrics=['accuracy'])

model.fit(X_train,Y_train, epochs=100 , batch_size=BATCH_SIZE 
           , verbose=True , validation_split=.2)

Train on 48000 samples, validate on 12000 samples
Epoch 1/100
48000/48000 [==============================] - 1s 30us/step - loss: 0.2853 - acc: 0.9159 - val_loss: 0.1931 - val_acc: 0.9426
Epoch 2/100
48000/48000 [==============================] - 1s 26us/step - loss: 0.2816 - acc: 0.9178 - val_loss: 0.1884 - val_acc: 0.9443
Epoch 3/100
48000/48000 [==============================] - 1s 26us/step - loss: 0.2751 - acc: 0.9201 - val_loss: 0.1849 - val_acc: 0.9456
Epoch 4/100
48000/48000 [==============================] - 1s 25us/step - loss: 0.2695 - acc: 0.9198 - val_loss: 0.1804 - val_acc: 0.9462
Epoch 5/100
48000/48000 [==============================] - 1s 26us/step - loss: 0.2631 - acc: 0.9234 - val_loss: 0.1769 - val_acc: 0.9478
Epoch 6/100
48000/48000 [==============================] - 1s 26us/step - loss: 0.2580 - acc: 0.9243 - val_loss: 0.1738 - val_acc: 0.9489
Epoch 7/100
48000/48000 [==============================] - 1s 26us/step - loss: 0.2511 - acc: 0.9259 - val_loss: 0.1708 - 

48000/48000 [==============================] - 1s 28us/step - loss: 0.1399 - acc: 0.9588 - val_loss: 0.1040 - val_acc: 0.9690
Epoch 60/100
48000/48000 [==============================] - 1s 30us/step - loss: 0.1380 - acc: 0.9592 - val_loss: 0.1042 - val_acc: 0.9680
Epoch 61/100
48000/48000 [==============================] - 1s 27us/step - loss: 0.1368 - acc: 0.9590 - val_loss: 0.1032 - val_acc: 0.9685
Epoch 62/100
48000/48000 [==============================] - 1s 27us/step - loss: 0.1361 - acc: 0.9584 - val_loss: 0.1025 - val_acc: 0.9695
Epoch 63/100
48000/48000 [==============================] - 1s 27us/step - loss: 0.1364 - acc: 0.9590 - val_loss: 0.1021 - val_acc: 0.9688
Epoch 64/100
48000/48000 [==============================] - 1s 26us/step - loss: 0.1331 - acc: 0.9604 - val_loss: 0.1019 - val_acc: 0.9685
Epoch 65/100
48000/48000 [==============================] - 1s 27us/step - loss: 0.1302 - acc: 0.9608 - val_loss: 0.1014 - val_acc: 0.9691
Epoch 66/100
48000/48000 [==============

In [36]:
# evaluations 
score = model.evaluate(X_test,Y_test)
print ("test score",score[0])
print ("test accuracy",score[1]*100)

10000/10000 [==============================] - 0s 26us/step
test score 0.08849719858358149
test accuracy 97.28999999999999


# trying different optimizers

use the previous model with differenet optimizers

In [40]:
# first i will use RMSprob
OPTIMIZER = RMSprop()
model.compile(loss='categorical_crossentropy',optimizer=OPTIMIZER,
              metrics=['accuracy'])
model.fit( X_train , Y_train , epochs=20 , 
          batch_size=BATCH_SIZE ,verbose=True , validation_split=.2)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 2s 34us/step - loss: 0.1548 - acc: 0.9564 - val_loss: 0.0974 - val_acc: 0.9746
Epoch 2/20
48000/48000 [==============================] - 1s 27us/step - loss: 0.1202 - acc: 0.9661 - val_loss: 0.0948 - val_acc: 0.9752
Epoch 3/20
48000/48000 [==============================] - 1s 27us/step - loss: 0.1131 - acc: 0.9681 - val_loss: 0.0925 - val_acc: 0.9753
Epoch 4/20
48000/48000 [==============================] - 1s 28us/step - loss: 0.1016 - acc: 0.9713 - val_loss: 0.0935 - val_acc: 0.9751
Epoch 5/20
48000/48000 [==============================] - 1s 28us/step - loss: 0.0970 - acc: 0.9712 - val_loss: 0.0913 - val_acc: 0.9772
Epoch 6/20
48000/48000 [==============================] - 1s 28us/step - loss: 0.0938 - acc: 0.9729 - val_loss: 0.0931 - val_acc: 0.9751
Epoch 7/20
48000/48000 [==============================] - 1s 30us/step - loss: 0.0867 - acc: 0.9751 - val_loss: 0.0933 - val_acc

In [41]:
score = model.evaluate(X_test,Y_test)
print ("test score ",score[0])
print ("test accuracy",score[1]*100)

10000/10000 [==============================] - 0s 24us/step
test score  0.07356165894430787
test accuracy 98.3


using RMSprob optimizer makes significant change in test accuracy

In [43]:
# sconde use Adam optimizer
OPTIMIZER = Adam()
model.compile(loss='categorical_crossentropy',optimizer=OPTIMIZER,
              metrics=['accuracy'])
model.fit( X_train , Y_train , epochs=20 , 
          batch_size=BATCH_SIZE ,verbose=True , validation_split=.2)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 2s 36us/step - loss: 0.0684 - acc: 0.9807 - val_loss: 0.0991 - val_acc: 0.9771
Epoch 2/20
48000/48000 [==============================] - 1s 28us/step - loss: 0.0627 - acc: 0.9818 - val_loss: 0.0928 - val_acc: 0.9780
Epoch 3/20
48000/48000 [==============================] - 1s 29us/step - loss: 0.0636 - acc: 0.9813 - val_loss: 0.0942 - val_acc: 0.9780
Epoch 4/20
48000/48000 [==============================] - 1s 29us/step - loss: 0.0545 - acc: 0.9834 - val_loss: 0.0869 - val_acc: 0.9787
Epoch 5/20
48000/48000 [==============================] - 1s 28us/step - loss: 0.0563 - acc: 0.9829 - val_loss: 0.0948 - val_acc: 0.9777
Epoch 6/20
48000/48000 [==============================] - 1s 29us/step - loss: 0.0558 - acc: 0.9823 - val_loss: 0.0910 - val_acc: 0.9775
Epoch 7/20
48000/48000 [==============================] - 1s 29us/step - loss: 0.0528 - acc: 0.9831 - val_loss: 0.0882 - val_acc

In [44]:
score = model.evaluate(X_test,Y_test)
print ("test score ",score[0])
print ("test accuracy",score[1]*100)

10000/10000 [==============================] - 0s 26us/step
test score  0.07380090379281683
test accuracy 98.21
